Unpack nibbles optimization

squiddy opened the pull request:
Hi,
first of all, thank you for your work. I'm interested in helping here, so I had a look at the code and played a bit with the server today. 

While playing around I noticed it uses quite a bit of CPU (at least on my pc), so I did a bit of profiling and it seems that unpack_nibbles in one case used 28% of the total processing time for filling up the cache when starting the server (it didn't change a lot when connecting with a client and waiting for all chunks to load). Using numpy I could reduce that significant, so in my tests it's now barely noticable. Other parts of the code stand out more. That's my first real encounter with numpy, so there probably is some space for further optimization.

I'm not sure if you can recreate this, I could provide profiling stats if you want. Please tell me if and how I can improve the description in case there are questions.

Regards,
Reiner

squiddy committed:
utilities: speed up nibble unpacking

MostAwesomeDude commented:
Nifty. Numpy is really not my forte. I'm NAKing this as-is because it breaks the corresponding unit test; please either fix the test or your changes.

squiddy commented:
I'm not sure what solution is more appropriate here. pack_nibbles() accepts an numpy array and returns a string, so I'd say it makes sense that unpack_nibbles() accepts strings only, and returns an array. Therefore I would update the test, but I'm not that comfortable with the code yet, so for all I know unpack_nibbles() might be called with a list somewhere/somewhen.

MostAwesomeDude commented:
git-grep is your friend. The only spots remaining where it is used are in bravo/serializers/alpha.py, where its output is immediately passed to np.array(). The only other spot it might be used is in Chunk, if a load_from_packet() from one of the clients is ever pulled in. In that case, it would also immediately be passed to np.array(). So I think returning an array would be appropriate.

On another note, using Python operators makes the code far more obvious; please do that as well.

squiddy committed:
Fixed the test case for unpack_nibbles

Also, unpack_nibbles will return an iterator for 1-dimensional view on
the array. That way, the assert_array_equal in the test works.
I adjusted the examples in the docstring and the testcase to not pass a
list anymore, instead it uses a string.

squiddy commented:
Sorry, I think I have phrased that wrong. I was talking about the input of the function. In test_unpack_nibbles() it's testing for both ["a"] and "nibbles". Using a list is one of the cases where the test fails, as numpy.fromstring doesn't seem to handle this well. That's why I was asking if the function ever gets passed a list, or byte strings only. After looking at the code I suppose it will be passed byte strings (as result of parsing NBT byte arrays). I was just confused about the test case using ["a"].

I agree that using Python operators makes the code more obvious, but as far as I know, using numpys functions is what probably makes the performance difference here. That way, it's probably working directly on the C array. I haven't used numpy much before, but I find that reasonable.

MostAwesomeDude commented:
Looks good. Didn't know about that numpy testing helper; I'll be using that on some other tests too. Merged.
